### ISSUE TITLE
`wrong_self_convention` wrongly detects `s: Self` as `self`

### ISSUE THREAD

**Issue by tspiteri** [rust-lang/rust-clippy#7179]
```rust
pub struct S(i32);

impl S {
    pub fn from_be(s: Self) -> Self {
        S(i32::from_be(s.0))
    }
}
```

Clippy says:
```
    Checking playground v0.0.1 (/playground)
warning: methods called `from_*` usually take no `self`
 --> src/lib.rs:4:20
  |
4 |     pub fn from_be(s: Self) -> Self {
  |                    ^
  |
  = note: `#[warn(clippy::wrong_self_convention)]` on by default
  = help: consider choosing a less ambiguous name
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention

warning: 1 warning emitted

    Finished dev [unoptimized + debuginfo] target(s) in 0.57s
```

### Meta

- `cargo clippy -V`: 0.1.53 (2021-05-05 bacf770)

**Comment by tspiteri** [comment]
> Well, apart from the name, those are equivalent, and you could use `self` instead of `s`.

They are equivalent in the implementation of the method, but not in calling the method. You cannot call `12i32.from_be()` (which is what the lint is protecting against) but you can call `i32::from_be(12i32)` (which should be allowed by the lint). As is the lint would not like the standard library `from_be` function for example.
